src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::credentials::Result;
+use crate::credentials::internal::sts_exchange::ClientAuthentication;
+use crate::headers_util::build_bearer_headers;
+use crate::token::{CachedTokenProvider, Token, TokenProvider};
+use crate::token_cache::TokenCache;
+use gax::error::CredentialsError;
+use http::{Extensions, HeaderMap};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::time::{Duration, Instant};
+
+use super::Credentials;
+use super::dynamic::CredentialsProvider;
+use super::external_account_sources::url_sourced_account::UrlSourcedCredentials;
+use super::internal::sts_exchange::{ExchangeTokenRequest, STSHandler};
+
+#[async_trait::async_trait]
+pub(crate) trait SubjectTokenProvider: std::fmt::Debug + Send + Sync {
+    /// Generate subject token that will be used on STS exchange.
+    async fn subject_token(&self) -> Result<String>;
```

Comment:
as a future optimization, we could use the cachableresource approach that @harkamaljot  is introducing. That way, if the subject token is not changed, we need not proceed with going to STS etc.